lib/elixir/src/elixir_quote.erl

```diff
@@ -165,12 +165,12 @@ do_escape(BitString, _) when is_bitstring(BitString) ->
 
 do_escape(Map, Q) when is_map(Map) ->
   TT =
-    [if
-      is_reference(V) ->
+    [case extract_value_ref(V) of
```

Comment:
Instead of extracting value, could we perhaps do a `do_quote_map_value` that checks for references, tuples, and otherwise just fallbacks to `do_quote`, returning either `{ok, Quoted}` or `{error, Ref}`?

Reply:
@josevalim is it what you meant? [30ee6fd](https://github.com/elixir-lang/elixir/pull/14560/commits/30ee6fda46432ebcc4f7cca0b0adc35e06c268ee)

I personally prefer the previous version since it's slightly more DRY and needs less params to be passed around, but no strong opinion.

I have another idea too, will try to push as an alternative proposal and let you chose 🙂

Reply:
I kinda like this one [26f3584](https://github.com/elixir-lang/elixir/pull/14560/commits/26f35841be9fe408ef463520cf9ec6980c801b31) readability-wise, even if line-wise it's slightly longer.
But I liked the initial version too.

Reply:
> @josevalim is it what you meant? [30ee6fd](https://github.com/elixir-lang/elixir/pull/14560/commits/30ee6fda46432ebcc4f7cca0b0adc35e06c268ee)

The idea is that you would already do the quoting as you traverse the tuple. :) This way it would be easier to traverse other data structures (even lists) without additional overhead. Although it does come with some duplication.

It is your call though! Pick whatever you prefer and ship it!